client/executor/wasmtime/src/instance_wrapper.rs

```diff
@@ -0,0 +1,245 @@
+//! runtime module.
+
+use crate::util;
+use sc_executor_common::error::{Error, Result};
+use sp_wasm_interface::{Pointer, WordSize};
+use std::slice;
+use wasmtime::{Instance, Memory, Table};
+
+/// Wrap the given WebAssembly Instance of a wasm module with Substrate-runtime.
+///
+/// This struct is a handy wrapper around a wasmtime `Instance` that provides substrate specific
+/// routines.
+pub struct InstanceWrapper {
+	instance: Instance,
+	memory: Memory,
+	table: Option<Table>,
+}
+
+impl InstanceWrapper {
+	/// Wrap the given `Instance`.
+	///
+	/// # Safety
+	///
+	/// This wrapper requires exclusive ownership over the given instance and all its components.
+	/// Since `Instance` is a `Clone` we cannot use typesystem to prevent this, hence unsafe.
+	///
+	/// Requiring `Instance` is rather a convenience feature: in reality exclusive ownership is only
+	/// required for `Memory` to guarantee exclusive access to its memory and prevent caching
+	/// the pointer to the memory's base address to prevent invalidation.
+	pub unsafe fn new(instance: Instance) -> Result<Self> {
+		Ok(Self {
+			table: get_table(&instance),
+			memory: get_linear_memory(&instance)?,
+			instance,
+		})
+	}
+
+	/// Increases the size of the linear memory attached to this instance.
+	pub fn grow_memory(&self, pages: u32) -> Result<()> {
+		if self.memory.grow(pages).is_ok() {
+			Ok(())
+		} else {
+			return Err("failed top increase the linear memory size".into());
+		}
+	}
+
+	/// Resolves a substrate entrypoint by the given name.
+	///
+	/// An entrypoint must have a signature `(i32, i32) -> i64`, otherwise this function will return
+	/// an error.
+	pub fn resolve_entrypoint(&self, name: &str) -> Result<wasmtime::Func> {
+		// Resolve the requested method and verify that it has a proper signature.
+		let export = self
+			.instance
+			.get_export(name)
+			.ok_or_else(|| Error::from(format!("Exported method {} is not found", name)))?;
+		let entrypoint = export
+			.func()
+			.ok_or_else(|| Error::from(format!("Export {} is not a function", name)))?;
+		match (entrypoint.ty().params(), entrypoint.ty().results()) {
+			(&[wasmtime::ValType::I32, wasmtime::ValType::I32], &[wasmtime::ValType::I64]) => {}
+			_ => {
+				return Err(Error::from(format!(
+					"method {} have an unsupported signature",
+					name
+				)))
+			}
+		}
+		Ok(entrypoint.clone())
+	}
+
+	/// Returns an indirect function table of this instance.
+	pub fn table(&self) -> Option<&Table> {
+		self.table.as_ref()
+	}
+
+	/// Returns the byte size of the linear memory instance attached to this instance.
+	pub fn memory_size(&self) -> u32 {
+		self.memory.data_size() as u32
+	}
+
+	/// Reads `__heap_base: i32` global variable and returns it.
+	///
+	/// If it doesn't exist, not a global or of not i32 type returns an error.
+	pub fn extract_heap_base(&self) -> Result<u32> {
+		let heap_base_export = self
+			.instance
+			.get_export("__heap_base")
```

Comment:
Should stuff like `__heap_base` rather be declared as some constants? Usually I prefer that over random strings sprinkled all over, at least you can maintain some convention, even if the constant is only used in one place.

Reply:
That would also give a place to leave some docs!

Reply:
I think it would be a good idea to address this in a follow-up PR!

Reply:
Fine with me!